Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,48 @@
+- **Troublesome programs** - Similar to what is already being logged in the heartbeat, look at which services or programs are responsible for hydrating large amounts of files in the enlistment
+
+### Potentially detrimental to access on a mounted enlistment
+
+- **Placeholder details** - Crawling the file system to enumerate directories of placeholders may unintentionally further hydrate files (IE downloading file sizes), but doing so would allow for greater details such as the percent hydration of folders, the total number of hydrated files, the type of files which are most hydrated, etc.
+- **File system** - Provide enumerated details for files in the placeholder list or the modified path list
+
+## Background Information
+
+- **Modified paths** is the list of files which VFS4G has completely pulled from the remote and whose contents are on disk in their entirety. These files are handed to git via a hook when any git command requiring the list of changed files is called
+- **Placeholders** are files which have been opened by the local system with a read tag but without a write tag. They are objects tracked by VFS4G since git should not have to internally scan them since it is guaranteed that they have not been modified and accordingly do not have to be rehashed for differences. Their contents are on disk, but control over them has not been entirely ceded over to the system. They decrease performance because VFS4G has to track all of them individually, and this can cause certain `gvfs` commands to be slow
+- The **Git index** is gits internal list of files which have been modified (tracked or untracked) which is what is used for commit commands to keep score of everything going on inside of the enlistment. It is useful to this project by means of comparing it to the list of **hydrated files** to deduce if those files should truly be hydrated or if it may have occurred incidentally.
+- The **file system** must be carefully considered in the case of working with a mounted repository. VFS4G intercepts file I/O calls at the kernel level to inject its own behavior (the projected file system and if needed a network call to pull down the file which the system asked for in the correct format). Because of this, enumerating directories can cause additional data to be pulled down from the remote by hydrating extra files or further hydrating those that already exist on disk. Accordingly, care must be taken when interacting with the file system lest the tool designed to help users increase enlistment heath end up causing more problems.
+
+## Components
+
+- **API** - The server component of the client-server relationship. Responsible for the logic and computation of statistical data, as well as the assembly and sending of a **json statistics object** over the named pipe to the client. It sits within the `InProgressMount` process and interacts with the VFS4G internals

```

Comment:
This **API** idea still worries me. Putting a new API in the named pipe layer still seems unnecessary (until we are forced to do it because we find a reason we can't access a resource from multiple processes). Instead, I'd start with the verb, and what work the verb needs to do in order to output a simple text-based message to a user summarizing what you found. Something like:

```
> gvfs statistics
Paths tracked by Git:   3,123,456 (100%)
Placeholders:      123,456 (4%)
Modified Paths:      1,234 (<1%)

Repo status: Healthy (no recommended action)

> (do stuff)

> gvfs statistics
Paths tracked by Git:   3,123,456 (100%)
Placeholders:      3,523,456 (111%)
Modified Paths:      500,000 (24%)

Repo status: Over-hydrated
Recommended Action: Run "gvfs dehydrate --confirm --no-status"
Take action? (y/n): y

Starting dehydration. All of your existing files will be backed up in C:\_git\os\dehydrate_backup\20190530_151645
WARNING: If you abort the dehydrate after this point, the repo may become corrupt

Unmounting...Succeeded
Authenticating...Succeeded
Backing up your files...Succeeded
Downloading git objects...Succeeded
Recreating git index...Succeeded
Mounting...Succeeded

Paths tracked by Git:   3,123,456 (100%)
Placeholders:      0 (0%)
Modified Paths:   0 (0%)
>
```
(Btw, having something like the input/output example above would be a good concrete example of something you could put in this document.)

You can take a look at `RepairVerb` for the concept of "repair jobs" which are different things that verb can do and report to the verb for details. Likely, you will want these different analyses to fit that sort of plug-in model.

Finally, you can still specify a JSON API layer, but as output from the _verb_, so your "additional visualization tools" could call the verb directly with a parameter (say, `--json`) and interpret the output nicely.

Reply:
Ironically as you were suggesting a console mock up Jameson was at my desk saying the same thing. I banged out an example locally but I will revisit it now with some of the ideas you had in yours as well as a loose example of what I will be building towards. As for the API, it does make sense to give a `--json` parameter as another method of passing data from the overall system. I'll take a look at the repair verb and consider that. I will continue working on plans for the system as a whole and investigate which statistics we can access readily from the verb vs. which ones may not be as accessible and might require the API.

Reply:
I like this!  I was thinking something similar but by folder so
```
DirectoryA - X file placeholders Y folder placeholder Z git files
DirectoryB - X1 file placeholders Y1 folder placeholder Z1 git files
etc...
```
And then you could run `gvfs statistics DirectoryA` to get the stats on that directory

Reply:
Definitely put Kevin's idea in the document, it's really good. Save it for a follow-up task after your initial prototype is complete.

Reply:
> This API idea still worries me. Putting a new API in the named pipe layer still seems unnecessary (until we are forced to do it because we find a reason we can't access a resource from multiple processes). Instead, I'd start with the verb, and what work the verb needs to do in order to output a simple text-based message to a user summarizing what you found.

I agree with @derrickstolee here.  @kewillford, do you know if the SQLite DB for placeholders will allow a second process to connect/read while VFS4G has open connection(s)?

Reply:
Yes SQLite allows for multiple connections from different processes.